File: medium/017_Letter_Combinations_of_a_Phone_Number/letter_combination.py
class Solution(object):
    digits_map = {"2":"abc",  "3":"def", "4":"ghi", "5":"jkl", "6":"mno",
                  "7":"pqrs", "8":"tuv", "9":"wxyz"}

    def letter_combination_internal(self, digits, path, result):
        depth = len(path)
        if depth == len(digits):
            result.append("".join(path))
            return

        for ch in self.digits_map[digits[depth]]:
            path.append(ch)
            self.letter_combination_internal(digits, path, result)
            path.pop()

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if len(digits) == 0:
            return []
        path, result = [], []
        self.letter_combination_internal(digits, path, result)
        return result

def test_combination(digits):
    result = solution.letterCombinations(digits)
    print "combination for {}".format(digits)
    for s in result:
        print s
    print

solution = Solution()
test_combination("23")
test_combination("234")

File: medium/017_Letter_Combinations_of_a_Phone_Number/letter_combination.go
package main

import "fmt"

var dMap = map[byte]string{
	'1': "",
	'2': "abc",
	'3': "def",
	'4': "ghi",
	'5': "jkl",
	'6': "mno",
	'7': "pqrs",
	'8': "tuv",
	'9': "wxyz",
}

func combInternal(digits string, results *[]string, seq []byte, idx int) {
	if idx >= len(digits) {
		*results = append(*results, string(seq))
		return
	}
	letters := dMap[digits[idx]]
	for i := 0; i < len(letters); i++ {
		seq[idx] = letters[i]
		combInternal(digits, results, seq, idx+1)
	}
}

func letterCombinations(digits string) []string {
	results := make([]string, 0)
	if len(digits) == 0 {
		return results
	}
	seq := make([]byte, len(digits))
	combInternal(digits, &results, seq, 0)
	return results
}

func testLetterCombinations(digits string) {
	results := letterCombinations(digits)
	fmt.Printf("digits %q, results %v\n", digits, results)
}

func main() {
	// output []
	testLetterCombinations("")
	// output ["a", "b", "c"]
	testLetterCombinations("2")
	// output ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"]
	testLetterCombinations("23")
}
